File: scripts/restore-household.py
#!/usr/bin/env python3
"""Restore Amplop household from Goodbudget CSV using Supabase REST API."""

import csv
import json
import math
import calendar
import os
import re
import ssl
import urllib.error
import urllib.request
from datetime import datetime

URL = "https://vqvknxpbdbqlibzhqutz.supabase.co/rest/v1"
KEY = None
OWNER = "stevenwilson614@gmail.com"
CSV_PATH = os.path.expanduser("~/Downloads/history.csv")

BASE_ENVELOPES = [
    ("Groceries", "Food", 0),
    ("Eating out", "Food", 1),
    ("Steven", "Personal", 2),
    ("Olivia", "Personal", 3),
    ("Private Classes", "Personal", 4),
    ("House Maintenance", "Other", 5),
    ("Rocky", "Other", 6),
    ("Vacation / fun", "Other", 7),
    ("Grab / Gas", "Other", 8),
    ("Giving", "Other", 9),
]

REMAININGS = {
    "Groceries": 4293414,
    "Eating Out": 1078590,
    "Steven": 1996513,
    "Olivia": 3844453,
    "Vacation / Fun": 36432461,
    "Grab / Gas": 5497822,
    "Giving": 6799906,
    "Rocky": 1322711,
    "House Mainetence": 2080874,
    "Private classes": 1741052,
}

SKIP = {"Fill Envelopes", "Fill from Available", "Fill from Unallocated"}


def load_key():
    env_path = os.path.expanduser("~/amplop/.env.local")
    key = None
    with open(env_path) as f:
        for line in f:
            if line.startswith("SUPABASE_SERVICE_ROLE_KEY="):
                key = line.strip().split("=", 1)[1]
    if not key:
        raise SystemExit("Missing SUPABASE_SERVICE_ROLE_KEY in .env.local")
    return key


def req(method, path, body=None, prefer=None, retries=5):
    headers = {
        "apikey": KEY,
        "Authorization": f"Bearer {KEY}",
        "Content-Type": "application/json",
    }
    if prefer:
        headers["Prefer"] = prefer
    data = json.dumps(body).encode() if body is not None else None
    ctx = ssl.create_default_context()
    last_err = None
    for attempt in range(retries):
        r = urllib.request.Request(f"{URL}{path}", data=data, headers=headers, method=method)
        try:
            with urllib.request.urlopen(r, context=ctx, timeout=120) as resp:
                raw = resp.read().decode()
                if not raw:
                    return None
                return json.loads(raw)
        except urllib.error.HTTPError as e:
            err = e.read().decode()
            raise RuntimeError(f"{method} {path}: {e.code} {err}") from e
        except OSError as e:
            last_err = e
            if attempt + 1 < retries:
                wait = min(2 ** attempt, 30)
                print(f"  retry {attempt + 1}/{retries - 1} after {wait}s ({e})")
                import time

                time.sleep(wait)
                continue
            raise
    raise last_err


def norm_env(raw: str) -> str:
    t = re.sub(r"\s+", " ", raw.strip())
    if not t:
        return ""
    if ":" in t:
        return t.split(":")[-1].strip()
    return t


def norm_date(raw: str):
    raw = raw.strip()
    if re.match(r"^\d{4}-\d{2}-\d{2}$", raw):
        return raw
    m = re.match(r"^(\d{1,2})/(\d{1,2})/(\d{4})$", raw)
    if m:
        return f"{m.group(3)}-{int(m.group(1)):02d}-{int(m.group(2)):02d}"
    try:
        return datetime.fromisoformat(raw).strftime("%Y-%m-%d")
    except ValueError:
        return None


def parse_amt(raw: str) -> int:
    cleaned = re.sub(r"[^\d.,-]", "", raw.strip())
    if not cleaned:
        return 0
    neg = cleaned.startswith("-")
    digits = cleaned.replace("-", "").replace(",", "")
    val = int(round(float(digits)))
    return -val if neg else val


def parse_csv(path):
    rows = []
    with open(path, newline="", encoding="utf-8-sig") as f:
        reader = csv.DictReader(f)
        for row in reader:
            date = norm_date(row.get("Date", ""))
            if not date:
                continue
            name = (row.get("Name") or "").strip()
            env_raw = (row.get("Envelope") or "").strip()
            amt = parse_amt(row.get("Amount") or "")
            if name in SKIP:
                continue
            if name == "Income":
                rows.append({"kind": "income", "date": date, "key": "", "name": name, "amt": abs(amt)})
                continue
            key = norm_env(env_raw)
            if amt < 0:
                rows.append({"kind": "expense", "date": date, "key": key, "name": name, "amt": abs(amt)})
            elif amt > 0 and key:
                rows.append({"kind": "transfer", "date": date, "key": key, "name": name, "amt": amt})
    return rows


def pair_transfers(rows):
    transfers = [r for r in rows if r["kind"] == "transfer"]
    used = set()
    pairs = []
    for i, a in enumerate(transfers):
        if i in used:
            continue
        for j in range(i + 1, len(transfers)):
            if j in used:
                continue
            b = transfers[j]
            if a["date"] != b["date"] or a["amt"] != b["amt"]:
                continue
            pairs.append({"date": a["date"], "from": a["key"], "to": b["key"], "amt": a["amt"]})
            used.add(i)
            used.add(j)
            break
    return pairs


def round_budget(n: float) -> int:
    if n <= 0:
        return 0
    if n < 1_000_000:
        return int(math.ceil(n / 10_000) * 10_000)
    return int(math.ceil(n / 100_000) * 100_000)


def compute_monthly_budgets(rows):
    now = datetime.now()
    year = now.year
    month = now.month - 5
    while month <= 0:
        month += 12
        year -= 1
    cutoff = datetime(year, month, 1)
    month_totals = {}
    month_mtd = {}
    current_month = now.strftime("%Y-%m")

    for row in rows:
        if row["kind"] != "expense" or not row["key"]:
            continue
        d = datetime.fromisoformat(row["date"])
        key = row["key"].lower()
        month_key = row["date"][:7]
        if d >= cutoff:
            month_totals.setdefault(key, {})
            month_totals[key][month_key] = month_totals[key].get(month_key, 0) + row["amt"]
        if month_key == current_month:
            month_mtd[key] = month_mtd.get(key, 0) + row["amt"]

    budgets = {}
    day = now.day
    days_in_month = calendar.monthrange(now.year, now.month)[1]
    for key, months in month_totals.items():
        vals = list(months.values())
        avg = sum(vals) / len(vals) if vals else 0
        mtd = month_mtd.get(key, 0)
        projected = round((mtd / day) * days_in_month) if day > 0 else mtd
        budgets[key] = round_budget(max(avg, projected, mtd))
    return budgets


def fetch_all(path):
    rows = []
    offset = 0
    page = 1000
    while True:
        chunk = req("GET", f"{path}&limit={page}&offset={offset}") or []
        rows.extend(chunk)
        if len(chunk) < page:
            break
        offset += page
    return rows


def find_env(envelopes, key):
    q = key.lower()
    for e in envelopes:
        if e["name"].lower() == q:
            return e
    for e in envelopes:
        if norm_env(e["name"]).lower() == q:
            return e
    for e in envelopes:
        n = e["name"].lower()
        if q in n or n in q:
            return e
    for alias, canonical in [("eating out", "eating out"), ("vacation / fun", "vacation / fun"), ("house mainetence", "house maintenance"), ("private classes", "private classes")]:
        if q == alias:
            for e in envelopes:
                if e["name"].lower() == canonical:
                    return e
    return None


def main():
    global KEY
    KEY = load_key()

    owner_rows = req("GET", f"/users?select=id,household_id,email&email=eq.{OWNER}")
    owner = owner_rows[0]
    hh = owner["household_id"]
    uid = owner["id"]
    print(f"Household {hh}")

    cat_ids = {}
    for i, name in enumerate(["Food", "Personal", "Other"]):
        existing = req("GET", f"/categories?select=id&household_id=eq.{hh}&name=eq.{name}&limit=1")
        if existing:
            cat_ids[name] = existing[0]["id"]
        else:
            created = req("POST", "/categories", {"household_id": hh, "name": name, "sort_order": i}, "return=representation")
            cat_ids[name] = created[0]["id"]

    envs = req("GET", f"/envelopes?select=*&household_id=eq.{hh}&trip_id=is.null") or []
    for name, cat, sort in BASE_ENVELOPES:
        hit = find_env(envs, name)
        if not hit:
            created = req(
                "POST",
                "/envelopes",
                {
                    "household_id": hh,
                    "name": name,
                    "category_id": cat_ids[cat],
                    "budget_amount": 0,
                    "budget_currency": "IDR",
                    "sort_order": sort,
                },
                "return=representation",
            )
            envs.append(created[0])
            print(f"Created {name}")
        else:
            req("PATCH", f"/envelopes?id=eq.{hit['id']}", {"category_id": cat_ids[cat], "sort_order": sort})

    def keep_tx(notes: str) -> bool:
        n = notes or ""
        return n.startswith("trip-draw:") or n.startswith("Goodbudget import")

    txs = req("GET", f"/transactions?select=id,notes&household_id=eq.{hh}") or []
    del_ids = [t["id"] for t in txs if not keep_tx(t.get("notes") or "")]
    for i in range(0, len(del_ids), 50):
        chunk = del_ids[i : i + 50]
        ids = ",".join(chunk)
        req("DELETE", f"/transaction_allocations?transaction_id=in.({ids})")
        req("DELETE", f"/transactions?id=in.({ids})")
    print(f"Deleted {len(del_ids)} transactions")

    existing = fetch_all(
        f"/transactions?select=date,merchant_name,amount,notes&household_id=eq.{hh}&notes=like.Goodbudget%20import%25",
    )
    existing_keys = {
        (t["date"], t.get("merchant_name") or "", int(t["amount"]))
        for t in existing
        if (t.get("notes") or "").startswith("Goodbudget import")
    }
    print(f"Already imported: {len(existing_keys)}")

    parsed = parse_csv(CSV_PATH)
    expenses = [r for r in parsed if r["kind"] == "expense"]
    transfers = pair_transfers(parsed)
    print(f"Importing {len(expenses)} expenses, {len(transfers)} transfers")

    imported = skipped = 0
    for row in expenses:
        key = (row["date"], row["name"], row["amt"])
        if key in existing_keys:
            continue
        env = find_env(envs, row["key"])
        if not env:
            skipped += 1
            continue
        tx = req(
            "POST",
            "/transactions",
            {
                "household_id": hh,
                "user_id": uid,
                "tx_type": "expense",
                "amount": row["amt"],
                "currency": "IDR",
                "amount_idr_snapshot": row["amt"],
                "fx_rate_snapshot": 1,
                "date": row["date"],
                "merchant_name": row["name"],
                "notes": f"Goodbudget import {row['date']}",
            },
            "return=representation",
        )[0]
        req(
            "POST",
            "/transaction_allocations",
            {"transaction_id": tx["id"], "envelope_id": env["id"], "amount": row["amt"]},
        )
        existing_keys.add(key)
        imported += 1
        if imported % 250 == 0:
            print(f"  {imported} new expenses...")

    transfer_keys = {
        (t["date"], int(t["amount"]))
        for t in existing
        if (t.get("notes") or "").startswith("Goodbudget import transfer")
    }
    tcount = 0
    for pair in transfers:
        tkey = (pair["date"], pair["amt"])
        if tkey in transfer_keys:
            continue
        fe = find_env(envs, pair["from"])
        te = find_env(envs, pair["to"])
        if not fe or not te:
            skipped += 1
            continue
        tx = req(
            "POST",
            "/transactions",
            {
                "household_id": hh,
                "user_id": uid,
                "tx_type": "transfer",
                "amount": pair["amt"],
                "currency": "IDR",
                "amount_idr_snapshot": pair["amt"],
                "fx_rate_snapshot": 1,
                "date": pair["date"],
                "notes": f"Goodbudget import transfer {pair['date']}",
            },
            "return=representation",
        )[0]
        req(
            "POST",
            "/transaction_allocations",
            [
                {"transaction_id": tx["id"], "envelope_id": fe["id"], "amount": pair["amt"]},
                {"transaction_id": tx["id"], "envelope_id": te["id"], "amount": -pair["amt"]},
            ],
        )
        transfer_keys.add(tkey)
        tcount += 1

    all_txs = fetch_all(
        f"/transactions?select=date,amount,amount_idr_snapshot,allocations:transaction_allocations(envelope_id,amount)&household_id=eq.{hh}",
    )
    spent = {}
    month_spent = {}
    carryover_month = datetime.now().strftime("%Y-%m")
    month_start = f"{carryover_month}-01"
    for t in all_txs:
        total = float(t.get("amount") or 0)
        total_idr = float(t.get("amount_idr_snapshot") or 0)
        if not total:
            continue
        d = t.get("date") or ""
        for a in t.get("allocations") or []:
            idr = round((float(a["amount"]) / total) * total_idr)
            spent[a["envelope_id"]] = spent.get(a["envelope_id"], 0) + idr
            if d >= month_start:
                month_spent[a["envelope_id"]] = month_spent.get(a["envelope_id"], 0) + idr

    csv_budgets = compute_monthly_budgets(parse_csv(CSV_PATH))

    budgets = 0
    for key, remaining in REMAININGS.items():
        env = find_env(envs, key)
        if not env:
            continue
        env_key = norm_env(env["name"]).lower()
        monthly = csv_budgets.get(env_key) or csv_budgets.get(key.lower()) or env.get("budget_amount") or 0
        ms = month_spent.get(env["id"], 0)
        carryover = remaining + ms - monthly
        req(
            "PATCH",
            f"/envelopes?id=eq.{env['id']}",
            {
                "budget_amount": monthly,
                "budget_currency": "IDR",
                "carryover_idr": carryover,
                "carryover_month": carryover_month,
            },
        )
        budgets += 1

    env_count = len(req("GET", f"/envelopes?select=id&household_id=eq.{hh}") or [])
    tx_count = len(req("GET", f"/transactions?select=id&household_id=eq.{hh}") or [])
    print(f"\nDone: {imported} expenses, {tcount} transfers, {skipped} skipped")
    print(f"Budgets updated: {budgets}")
    print(f"Totals: {env_count} envelopes, {tx_count} transactions")
    print("Refresh https://stevenwilson614.github.io/amplop/#/envelopes")


if __name__ == "__main__":
    main()

File: scripts/restore-household.mjs
#!/usr/bin/env node
/**
 * Restore household envelopes + Goodbudget transaction history.
 *
 * Usage:
 *   node scripts/restore-household.mjs [--csv /path/to/history.csv] [--owner stevenwilson614@gmail.com]
 */

import { createClient } from "@supabase/supabase-js";
import fs from "fs";
import path from "path";
import { fileURLToPath } from "url";

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const root = path.join(__dirname, "..");

const BASE_ENVELOPES = [
  { name: "Groceries", category: "Food", sort: 0 },
  { name: "Eating out", category: "Food", sort: 1 },
  { name: "Steven", category: "Personal", sort: 2 },
  { name: "Olivia", category: "Personal", sort: 3 },
  { name: "Private Classes", category: "Personal", sort: 4 },
  { name: "House Maintenance", category: "Other", sort: 5 },
  { name: "Rocky", category: "Other", sort: 6 },
  { name: "Vacation / fun", category: "Other", sort: 7 },
  { name: "Grab / Gas", category: "Other", sort: 8 },
  { name: "Giving", category: "Other", sort: 9 },
];

const REMAININGS = {
  Groceries: 4293414,
  "Eating Out": 1078590,
  Steven: 1996513,
  Olivia: 3844453,
  "Vacation / Fun": 36432461,
  "Grab / Gas": 5497822,
  Giving: 6799906,
  Rocky: 1322711,
  "House Mainetence": 2080874,
  "Private classes": 1741052,
};

function loadEnv() {
  const envPath = path.join(root, ".env.local");
  const env = {};
  for (const line of fs.readFileSync(envPath, "utf8").split("\n")) {
    if (!line || line.startsWith("#")) continue;
    const idx = line.indexOf("=");
    if (idx < 0) continue;
    env[line.slice(0, idx).trim()] = line.slice(idx + 1).trim();
  }
  return env;
}

function parseArgs(argv) {
  let csv = path.join(process.env.HOME ?? "", "Downloads/history.csv");
  let owner = "stevenwilson614@gmail.com";
  for (let i = 2; i < argv.length; i++) {
    if (argv[i] === "--csv") csv = argv[++i] ?? csv;
    else if (argv[i] === "--owner") owner = argv[++i] ?? owner;
  }
  return { csv, owner };
}

function splitCsvLine(line) {
  const out = [];
  let cur = "";
  let inQuotes = false;
  for (const ch of line) {
    if (ch === '"') { inQuotes = !inQuotes; continue; }
    if (ch === "," && !inQuotes) { out.push(cur); cur = ""; continue; }
    cur += ch;
  }
  out.push(cur);
  return out;
}

function normalizeDate(raw) {
  if (/^\d{4}-\d{2}-\d{2}$/.test(raw)) return raw;
  const m = raw.match(/^(\d{1,2})\/(\d{1,2})\/(\d{4})$/);
  if (m) return `${m[3]}-${m[1].padStart(2, "0")}-${m[2].padStart(2, "0")}`;
  const d = new Date(raw);
  if (isNaN(d.getTime())) return null;
  return d.toLocaleDateString("en-CA");
}

function normEnv(raw) {
  const t = raw.trim().replace(/\s+/g, " ");
  if (!t) return "";
  if (t.includes(":")) return t.split(":").map((p) => p.trim()).filter(Boolean).pop();
  return t;
}

function parseAmt(raw) {
  const cleaned = raw.trim().replace(/[^\d.,-]/g, "");
  if (!cleaned) return 0;
  const neg = cleaned.startsWith("-");
  const digits = cleaned.replace(/-/g, "").replace(/,/g, "");
  const v = Math.round(parseFloat(digits));
  return neg ? -v : v;
}

function parseGoodbudget(csvText) {
  const lines = csvText.trim().split(/\r?\n/).filter(Boolean);
  const header = splitCsvLine(lines[0]).map((h) => h.trim().toLowerCase());
  const dateIdx = header.indexOf("date");
  const envIdx = header.indexOf("envelope");
  const nameIdx = header.indexOf("name");
  const amtIdx = header.indexOf("amount");
  if (dateIdx < 0 || envIdx < 0 || nameIdx < 0 || amtIdx < 0) {
    throw new Error("Not a Goodbudget CSV");
  }

  const skip = new Set(["Fill Envelopes", "Fill from Available", "Fill from Unallocated"]);
  const rows = [];

  for (let i = 1; i < lines.length; i++) {
    const cols = splitCsvLine(lines[i]);
    const date = normalizeDate(cols[dateIdx]?.trim() ?? "");
    const envelopeRaw = cols[envIdx]?.trim() ?? "";
    const name = cols[nameIdx]?.trim() ?? "";
    const amountMinor = parseAmt(cols[amtIdx] ?? "");
    if (!date) continue;

    if (name === "Income") {
      rows.push({ date, envelopeKey: "", name, amountMinor: Math.abs(amountMinor), kind: "income" });
      continue;
    }
    if (skip.has(name)) {
      rows.push({ date, envelopeKey: "", name, amountMinor, kind: "skip" });
      continue;
    }

    const envelopeKey = normEnv(envelopeRaw);
    if (amountMinor < 0) {
      rows.push({ date, envelopeKey, name, amountMinor: Math.abs(amountMinor), kind: "expense" });
    } else if (amountMinor > 0 && envelopeKey) {
      rows.push({ date, envelopeKey, name, amountMinor, kind: "transfer" });
    }
  }
  return rows;
}

function pairTransfers(rows) {
  const transfers = rows.filter((r) => r.kind === "transfer");
  const used = new Set();
  const pairs = [];
  for (let i = 0; i < transfers.length; i++) {
    if (used.has(i)) continue;
    for (let j = i + 1; j < transfers.length; j++) {
      if (used.has(j)) continue;
      const a = transfers[i], b = transfers[j];
      if (a.date !== b.date || Math.abs(a.amountMinor) !== Math.abs(b.amountMinor) || a.amountMinor === b.amountMinor) continue;
      const from = a.amountMinor < 0 ? a : b;
      const to = a.amountMinor > 0 ? a : b;
      pairs.push({ date: a.date, fromKey: from.envelopeKey, toKey: to.envelopeKey, amountMinor: Math.abs(a.amountMinor) });
      used.add(i); used.add(j);
      break;
    }
  }
  return pairs;
}

function findEnvelope(envelopes, key) {
  const q = key.toLowerCase();
  return envelopes.find((e) => e.name.toLowerCase() === q)
    ?? envelopes.find((e) => normEnv(e.name).toLowerCase() === q)
    ?? envelopes.find((e) => e.name.toLowerCase().includes(q) || q.includes(e.name.toLowerCase()));
}

function monthsElapsed(createdAt) {
  const created = new Date(createdAt);
  const now = new Date();
  return Math.max(1, (now.getFullYear() - created.getFullYear()) * 12 + (now.getMonth() - created.getMonth()) + 1);
}

async function deleteTransactions(sb, householdId) {
  const { data: txs } = await sb
    .from("transactions")
    .select("id, notes")
    .eq("household_id", householdId);

  const ids = (txs ?? [])
    .filter((t) => !(t.notes ?? "").startsWith("trip-draw:"))
    .map((t) => t.id);

  for (let i = 0; i < ids.length; i += 50) {
    const chunk = ids.slice(i, i + 50);
    await sb.from("transaction_allocations").delete().in("transaction_id", chunk);
    await sb.from("transactions").delete().in("id", chunk);
  }
  return ids.length;
}

async function main() {
  const { csv, owner } = parseArgs(process.argv);
  if (!fs.existsSync(csv)) throw new Error(`CSV not found: ${csv}`);

  const env = loadEnv();
  const sb = createClient(env.VITE_SUPABASE_URL, env.SUPABASE_SERVICE_ROLE_KEY, {
    auth: { autoRefreshToken: false, persistSession: false },
  });

  const { data: ownerRow, error: ownerErr } = await sb
    .from("users")
    .select("id, household_id, email")
    .ilike("email", owner)
    .maybeSingle();
  if (ownerErr) throw ownerErr;
  if (!ownerRow) throw new Error(`No user row for ${owner}`);

  const householdId = ownerRow.household_id;
  const userId = ownerRow.id;
  console.log(`Household ${householdId} (owner ${ownerRow.email})`);

  // Ensure categories
  const categoryNames = ["Food", "Personal", "Other"];
  const categoryIds = {};
  for (const name of categoryNames) {
    const { data: existing } = await sb
      .from("categories")
      .select("id")
      .eq("household_id", householdId)
      .ilike("name", name)
      .limit(1)
      .maybeSingle();
    if (existing) {
      categoryIds[name] = existing.id;
      continue;
    }
    const { data: created, error } = await sb
      .from("categories")
      .insert({ household_id: householdId, name, sort_order: Object.keys(categoryIds).length })
      .select("id")
      .single();
    if (error) throw error;
    categoryIds[name] = created.id;
  }

  // Ensure base envelopes (household only)
  const { data: existingEnvs } = await sb
    .from("envelopes")
    .select("*")
    .eq("household_id", householdId)
    .is("trip_id", null);

  const householdEnvs = [...(existingEnvs ?? [])];
  for (const spec of BASE_ENVELOPES) {
    let env = findEnvelope(householdEnvs, spec.name);
    if (!env) {
      const { data: created, error } = await sb
        .from("envelopes")
        .insert({
          household_id: householdId,
          name: spec.name,
          category_id: categoryIds[spec.category],
          budget_amount: 0,
          budget_currency: "IDR",
          sort_order: spec.sort,
        })
        .select("*")
        .single();
      if (error) throw error;
      env = created;
      householdEnvs.push(created);
      console.log(`Created envelope: ${spec.name}`);
    } else {
      await sb.from("envelopes").update({
        category_id: categoryIds[spec.category],
        sort_order: spec.sort,
      }).eq("id", env.id);
    }
  }

  const deleted = await deleteTransactions(sb, householdId);
  console.log(`Deleted ${deleted} non-trip transactions`);

  const csvText = fs.readFileSync(csv, "utf8");
  const rows = parseGoodbudget(csvText);
  const expenses = rows.filter((r) => r.kind === "expense");
  const transfers = pairTransfers(rows);
  console.log(`Parsed ${expenses.length} expenses, ${transfers.length} transfer pairs`);

  const { data: envsForImport } = await sb
    .from("envelopes")
    .select("*")
    .eq("household_id", householdId)
    .is("trip_id", null);
  const envelopes = envsForImport ?? [];

  let imported = 0;
  let skipped = 0;
  for (const row of expenses) {
    const env = findEnvelope(envelopes, row.envelopeKey);
    if (!env) { skipped++; continue; }
    const { data: tx, error: txErr } = await sb.from("transactions").insert({
      household_id: householdId,
      user_id: userId,
      tx_type: "expense",
      amount: row.amountMinor,
      currency: "IDR",
      amount_idr_snapshot: row.amountMinor,
      fx_rate_snapshot: 1,
      date: row.date,
      merchant_name: row.name,
      notes: `Goodbudget import ${row.date}`,
    }).select("id").single();
    if (txErr || !tx) { skipped++; continue; }
    await sb.from("transaction_allocations").insert({
      transaction_id: tx.id,
      envelope_id: env.id,
      amount: row.amountMinor,
    });
    imported++;
    if (imported % 500 === 0) console.log(`  ${imported} expenses...`);
  }

  let transferCount = 0;
  for (const pair of transfers) {
    const fromEnv = findEnvelope(envelopes, pair.fromKey);
    const toEnv = findEnvelope(envelopes, pair.toKey);
    if (!fromEnv || !toEnv) { skipped++; continue; }
    const { data: tx } = await sb.from("transactions").insert({
      household_id: householdId,
      user_id: userId,
      tx_type: "transfer",
      amount: pair.amountMinor,
      currency: "IDR",
      amount_idr_snapshot: pair.amountMinor,
      fx_rate_snapshot: 1,
      date: pair.date,
      notes: `Goodbudget import transfer ${pair.date}`,
    }).select("id").single();
    if (!tx) { skipped++; continue; }
    await sb.from("transaction_allocations").insert([
      { transaction_id: tx.id, envelope_id: fromEnv.id, amount: pair.amountMinor },
      { transaction_id: tx.id, envelope_id: toEnv.id, amount: -pair.amountMinor },
    ]);
    transferCount++;
  }

  // Sync remainings → monthly budgets
  const { data: allTxs } = await sb
    .from("transactions")
    .select("amount, amount_idr_snapshot, allocations:transaction_allocations(envelope_id, amount)")
    .eq("household_id", householdId);

  const spentMap = {};
  for (const t of allTxs ?? []) {
    const total = Number(t.amount) || 0;
    const totalIdr = Number(t.amount_idr_snapshot) || 0;
    if (!total || !t.allocations) continue;
    for (const a of t.allocations) {
      const idr = Math.round((Number(a.amount) / total) * totalIdr);
      spentMap[a.envelope_id] = (spentMap[a.envelope_id] ?? 0) + idr;
    }
  }

  let budgetUpdates = 0;
  for (const [key, remaining] of Object.entries(REMAININGS)) {
    const env = findEnvelope(envelopes, key);
    if (!env) continue;
    const spent = spentMap[env.id] ?? 0;
    const months = monthsElapsed(env.created_at);
    const monthlyBudget = Math.max(0, Math.ceil((remaining + spent) / months));
    const { error } = await sb.from("envelopes").update({
      budget_amount: monthlyBudget,
      budget_currency: "IDR",
    }).eq("id", env.id);
    if (!error) budgetUpdates++;
  }

  const { count: txCount } = await sb
    .from("transactions")
    .select("id", { count: "exact", head: true })
    .eq("household_id", householdId);
  const { count: envCount } = await sb
    .from("envelopes")
    .select("id", { count: "exact", head: true })
    .eq("household_id", householdId);

  console.log(`\nRestore complete:`);
  console.log(`  imported ${imported} expenses, ${transferCount} transfers (${skipped} skipped)`);
  console.log(`  updated ${budgetUpdates} envelope budgets`);
  console.log(`  totals now: ${envCount} envelopes, ${txCount} transactions`);
  console.log(`\nRefresh https://stevenwilson614.github.io/amplop/#/envelopes`);
}

main().catch((err) => {
  console.error(err.message ?? err);
  process.exit(1);
});
